File: src/main/java/sigmaone/industrialism/block/wiring/BlockWireNode.java
package sigmaone.industrialism.block.wiring;

import net.minecraft.block.*;
import net.minecraft.block.entity.BlockEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.ItemPlacementContext;
import net.minecraft.item.Items;
import net.minecraft.state.StateManager;
import net.minecraft.state.property.IntProperty;
import net.minecraft.state.property.Properties;
import net.minecraft.text.LiteralText;
import net.minecraft.text.TranslatableText;
import net.minecraft.util.ActionResult;
import net.minecraft.util.Hand;
import net.minecraft.util.hit.BlockHitResult;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import net.minecraft.util.shape.VoxelShape;
import net.minecraft.util.shape.VoxelShapes;
import net.minecraft.world.BlockView;
import net.minecraft.world.World;
import sigmaone.industrialism.Industrialism;

public class BlockWireNode extends FacingBlock {
    public static final IntProperty STATE = IntProperty.of("state", 0, 2);
    public BlockWireNode(AbstractBlock.Settings settings) {
        super(settings.nonOpaque());
        this.setDefaultState(this.getStateManager().getDefaultState().with(STATE, 0).with(FACING, Direction.DOWN));
    }

    @Override
    protected void appendProperties(StateManager.Builder<Block, BlockState> stateBuilder) {
        stateBuilder.add(STATE);
        stateBuilder.add(Properties.FACING);
    }

    @SuppressWarnings("deprecation")
    @Override
    public VoxelShape getOutlineShape(BlockState state, BlockView world, BlockPos pos, ShapeContext context) {
        switch (state.get(FACING)) {
            case UP   : return VoxelShapes.cuboid((1f/16f)*6f, 0f, (1f/16f)*6f, (1f/16f)*10f, (1f/16f)*4f, (1f/16f)*10f);
            case DOWN : return VoxelShapes.cuboid((1f/16f)*6f, (1f/16f)*12f, (1f/16f)*6f, (1f/16f)*10f, (1f/16f)*16f, (1f/16f)*10f);
            case NORTH: return VoxelShapes.cuboid((1f/16f)*6f, (1f/16f)*6f, (1f/16f)*12f, (1f/16f)*10f, (1f/16f)*10f, 1f);
            case SOUTH: return VoxelShapes.cuboid((1f/16f)*6f, (1f/16f)*6f, 0f, (1f/16f)*10f, (1f/16f)*10f, (1f/16f)*4f);
            case EAST : return VoxelShapes.cuboid(0f, (1f/16f)*6f, (1f/16f)*6f, (1f/16f)*4f, (1f/16f)*10f, (1f/16f)*10f);
            case WEST : return VoxelShapes.cuboid(1f, (1f/16f)*6f, (1f/16f)*6f, (1f/16f)*12f, (1f/16f)*10f, (1f/16f)*10f);
            default: throw new IllegalStateException("Illegal side: " + state.get(FACING));
        }
    }

    @SuppressWarnings("deprecation")
    @Override
    public VoxelShape getCollisionShape(BlockState state, BlockView world, BlockPos pos, ShapeContext context) {
        return this.getOutlineShape(state, world, pos, context);
    }

    public BlockState getPlacementState(ItemPlacementContext context) {
        return this.getDefaultState().with(FACING, context.getSide());
    }

    @Override
    public void onBreak(World world, BlockPos pos, BlockState state, PlayerEntity player) {
        BlockEntityWireNode blockEntity = (BlockEntityWireNode) world.getBlockEntity(pos);
        blockEntity.removeAllConnections();
        super.onBreak(world, pos, state, player);
    }

    @SuppressWarnings("deprecation")
    @Override
    public ActionResult onUse(BlockState state, World world, BlockPos pos, PlayerEntity player, Hand hand, BlockHitResult hit) {
        BlockEntity blockEntity = world.getBlockEntity(pos);
        if (blockEntity == null) {
            return ActionResult.PASS;
        }

        if (!world.isClient) {
            if (blockEntity instanceof BlockEntityWireNode) {
                if (player.getStackInHand(hand).getItem() == Industrialism.SCREWDRIVER) {
                    String modeTranslated;

                    ((BlockEntityWireNode) blockEntity).cycleSideConfig(state.get(FACING).getOpposite());
                    ((BlockEntityWireNode) blockEntity).cycleIOState();
                    world.setBlockState(pos, state.with(STATE, ((BlockEntityWireNode) blockEntity).getIOState().ordinal()));

                    modeTranslated = new TranslatableText("variable."+Industrialism.MOD_ID+".ioconfig."+((BlockEntityWireNode) blockEntity)
                            .getSideConfig(state.get(FACING).getOpposite())
                            .toString().toLowerCase()).getString();
                    player.sendMessage(new TranslatableText("popup."+Industrialism.MOD_ID+".ioconfig.set_noside", modeTranslated), true);

                    return ActionResult.SUCCESS;
                }
            }
        }
        return ActionResult.PASS;
    }
}

File: src/main/kotlin/sigmaone/industrialism/block/wiring/BlockWireNode.kt
package sigmaone.industrialism.block.wiring

import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.block.BlockWithEntity
import net.minecraft.block.ShapeContext
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.ItemPlacementContext
import net.minecraft.state.StateManager
import net.minecraft.state.property.Properties
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.util.shape.VoxelShape
import net.minecraft.util.shape.VoxelShapes
import net.minecraft.world.BlockView
import net.minecraft.world.World
import sigmaone.industrialism.component.wiring.IComponentWireNode

abstract class BlockWireNode(settings: Settings, val height: Int) : BlockWithEntity(settings.nonOpaque()) {

    override fun appendProperties(stateBuilder: StateManager.Builder<Block, BlockState>) {
        stateBuilder.add(sigmaone.industrialism.util.Properties.IO)
        stateBuilder.add(Properties.FACING)
    }

    override fun getOutlineShape(state: BlockState, world: BlockView, pos: BlockPos, context: ShapeContext): VoxelShape {
        val f116 = 1.0/16.0
        val midMin = f116*6
        val midMax = f116*10
        val heightD = f116*height
        return when (state.get(Properties.FACING)) {
            Direction.UP    -> VoxelShapes.cuboid(midMin, 0.0, midMin, midMax, heightD, midMax)
            Direction.DOWN  -> VoxelShapes.cuboid(midMin, 1.0-heightD, midMin, midMax, 1.0, midMax)
            Direction.NORTH -> VoxelShapes.cuboid(midMin, midMin, 1.0-heightD, midMax, midMax, 1.0)
            Direction.SOUTH -> VoxelShapes.cuboid(midMin, midMin, 0.0, midMax, midMax, heightD)
            Direction.EAST  -> VoxelShapes.cuboid(0.0, midMin, midMin, heightD, midMax, midMax)
            Direction.WEST  -> VoxelShapes.cuboid(1.0-heightD, midMin, midMin, 1.0, midMax, midMax)
            else -> throw IllegalStateException("Illegal side: " + state.get(Properties.FACING))
        }
    }

    override fun getCollisionShape(state: BlockState, world: BlockView, pos: BlockPos, context: ShapeContext): VoxelShape {
        return getOutlineShape(state, world, pos, context)
    }

    override fun getPlacementState(context: ItemPlacementContext): BlockState? {
        return defaultState.with(Properties.FACING, context.side)
    }

    override fun onBreak(world: World, pos: BlockPos, state: BlockState, player: PlayerEntity) {
        val blockEntity = world.getBlockEntity(pos)
        if (blockEntity is IComponentWireNode<*>) {
            blockEntity.componentWireNode.removeAllConnections()
        }
        super.onBreak(world, pos, state, player)
    }

    init {
        defaultState = getStateManager()
            .defaultState.with(sigmaone.industrialism.util.Properties.IO, 0)
            .with(Properties.FACING, Direction.DOWN)
    }
}